File: examples/stochastic-policy-small/stochastic-policy.R
#!/usr/bin/env Rscript

# Small synthetic stochastic-policy workload.
#
# Model and data have been generated specifically for this example. Python
# and R scripts read the same MDL and CSV files so to make their results comparable.

suppressPackageStartupMessages(library(bimets))

script_arg <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
script_path <- normalizePath(sub("^--file=", "", script_arg[[1]]))
example_dir <- dirname(script_path)
model_path <- file.path(example_dir, "stochastic-policy.mdl")
data_path <- file.path(example_dir, "stochastic-policy-data.csv")

model_data <- CSV2BIMETS(data_path, mergedList = TRUE)
model_text <- paste(readLines(model_path, warn = FALSE), collapse = "
")

metric <- function(name, value) cat(sprintf("%s,%.6f
", name, value))

simulation_range <- c(2010, 1, 2011, 4)
base_model <- LOAD_MODEL(modelText = model_text, quietly = TRUE)
base_model <- LOAD_MODEL_DATA(base_model, model_data, quietly = TRUE)

stochastic <- STOCHSIMULATE(
  base_model,
  TSRANGE = simulation_range,
  simAlgo = "NEWTON",
  simConvergence = 1e-7,
  simIterLimit = 100,
  StochStructure = list(
    y001 = list(TSRANGE = TRUE, TYPE = "NORM", PARS = c(0, 0.05)),
    x001 = list(TSRANGE = TRUE, TYPE = "UNIF", PARS = c(-0.10, 0.10))
  ),
  StochReplica = 100,
  StochSeed = 123,
  quietly = TRUE
)
multipliers <- MULTMATRIX(
  base_model,
  TSRANGE = simulation_range,
  simAlgo = "NEWTON",
  simConvergence = 1e-7,
  simIterLimit = 100,
  TARGET = c("aggregate", "y001"),
  INSTRUMENT = c("x001", "y001"),
  quietly = TRUE
)
baseline <- SIMULATE(
  base_model,
  TSRANGE = simulation_range,
  simAlgo = "NEWTON",
  simConvergence = 1e-7,
  simIterLimit = 100,
  quietly = TRUE
)
desired <- baseline$simulation$aggregate * 1.01
targeting <- RENORM(
  base_model,
  TSRANGE = simulation_range,
  simAlgo = "NEWTON",
  simConvergence = 1e-7,
  simIterLimit = 100,
  TARGET = list(aggregate = desired),
  INSTRUMENT = "x001",
  renormIterLimit = 6,
  renormConvergence = 1e-6,
  quietly = TRUE
)
optimum <- OPTIMIZE(
  base_model,
  TSRANGE = simulation_range,
  simAlgo = "NEWTON",
  simConvergence = 1e-7,
  simIterLimit = 100,
  OptimizeBounds = list(x001 = list(TSRANGE = TRUE, BOUNDS = c(0.5, 1.5))),
  OptimizeRestrictions = list(r = list(
    TSRANGE = TRUE, INEQUALITY = "x001 >= 0.52 & x001 <= 1.48"
  )),
  OptimizeFunctions = list(f = list(
    TSRANGE = TRUE, FUNCTION = "aggregate-0.01*x001^2"
  )),
  StochReplica = 100,
  StochSeed = 321,
  quietly = TRUE
)
metric("equations", length(base_model$vendog))
metric("stochastic_mean_checksum", sum(stochastic$stochastic_simulation$aggregate$mean))
metric("stochastic_sd_checksum", sum(stochastic$stochastic_simulation$aggregate$sd))
metric("multiplier_checksum", sum(multipliers$MultiplierMatrix))
metric("renorm_instrument_checksum", sum(targeting$renorm$INSTRUMENT$x001))
metric("optimum", optimum$optimize$optFunMax)

File: examples/stochastic-policy-small/stochastic-policy.py
#!/usr/bin/env python3
"""Small synthetic stochastic-policy workload.

Model and data have been generated specifically for this example. Python
and R scripts read the same MDL and CSV files so to make their results comparable.
"""

from pathlib import Path

from bimets import (
    CSV2BIMETS,
    LOAD_MODEL,
    MULTMATRIX,
    OPTIMIZE,
    RENORM,
    SIMULATE,
    STOCHSIMULATE,
    BimetsDataset,
    StochasticDisturbance,
)

EXAMPLE_DIR = Path(__file__).resolve().parent
MODEL_PATH = EXAMPLE_DIR / "stochastic-policy.mdl"
DATA_PATH = EXAMPLE_DIR / "stochastic-policy-data.csv"


def load_data() -> BimetsDataset:
    """Load the shared quarterly CSV as a BIMETS dataset."""
    return BimetsDataset(CSV2BIMETS(DATA_PATH, merged=True))


def metric(name: str, value: float) -> None:
    """Print a stable numerical metric."""
    print(f"{name},{value:.6f}")


SIMULATION_RANGE = (2010, 1, 2011, 4)
REPLICAS = 100


def main() -> None:
    """Run stochastic simulation, multipliers, targeting, and optimal control."""
    model = LOAD_MODEL(model_text=MODEL_PATH.read_text(encoding="utf-8"))
    data = load_data()
    common = dict(
        coefficients={},
        time_range=SIMULATION_RANGE,
        algorithm="NEWTON",
        convergence=1e-7,
        max_iterations=100,
    )

    stochastic = STOCHSIMULATE(
        model,
        data,
        disturbances={
            "y001": StochasticDisturbance("NORMAL", (0.0, 0.05)),
            "x001": StochasticDisturbance("UNIFORM", (-0.10, 0.10)),
        },
        replicas=REPLICAS,
        seed=123,
        **common,
    )
    multipliers = MULTMATRIX(
        model,
        data,
        targets=("aggregate", "y001"),
        instruments=("x001", "y001"),
        **common,
    )
    baseline = SIMULATE(model, data, **common)
    desired = baseline["aggregate"] * 1.01
    targeting = RENORM(
        model,
        data,
        targets={"aggregate": desired},
        instruments="x001",
        renormalization_iterations=6,
        renormalization_convergence=1e-6,
        **common,
    )
    optimum = OPTIMIZE(
        model,
        data,
        bounds={"x001": (0.5, 1.5)},
        restrictions="x001 >= 0.52 & x001 <= 1.48",
        objective_functions="aggregate-0.01*x001^2",
        replicas=REPLICAS,
        seed=321,
        **common,
    )
    metric("equations", len(model.endogenous))
    metric("stochastic_mean_checksum", stochastic["aggregate"].mean.values.sum())
    metric("stochastic_sd_checksum", stochastic["aggregate"].sd.values.sum())
    metric("multiplier_checksum", multipliers.matrix.sum())
    metric("renorm_instrument_checksum", targeting.instruments["x001"].values.sum())
    metric("optimum", float(optimum.objective_max))


if __name__ == "__main__":
    main()
